### PR TITLE
Use EESSI module in EESSI-install-software.sh

### PR DESCRIPTION
**Pull request by casparvl**

### CODE REVIEW

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2

**Comment by casparvl** [comment]
Ghehe, is that a polite way of saying you would like a review? ;-)

I don't think it will outright affect the use in `EESSI-install-software.sh`, since we don't unload it here. But I'll put it on my radar - I want to review the cuDNN one first though...

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2

**Comment by casparvl** [comment]
```
Lmod Warning: You cannot use EESSI_CVMFS_INSTALL in combination with any other
EESSI_*_INSTALL environment variables
```
I guess I need to unset

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023b.yml [hunk 1]
```diff
@@ -7,3 +7,4 @@ easyconfigs:
       options:
         # see https://github.com/easybuilders/easybuild-easyconfigs/pull/21430
         from-commit: 8b509882d03402e2998ff9b22c154a6957e36d6b
+  - patchelf-0.18.0-GCCcore-13.2.0.eb:
```
Comment (boegel) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023b.yml:10]: Please use EasyBuild v4.9.4 instead, and also pull in the updated `patchelf` from https://github.com/easybuilders/easybuild-easyconfigs/pull/21674

Comment (casparvl) [reply easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023b.yml:10]: no worries, this is just for proving the `EESSI-install-software.sh` still works, it's just a test build. I'll strip it out before marking this ready for review, as I don't want it to be deployed ;)

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2

**Comment by casparvl** [comment]
Hmm...

```
size: 0 MiB (3659 bytes)
entries: 1
modules under 2023.06/software/linux/x86_64/amd/zen2/modules/all
no module files in tarball
software under 2023.06/software/linux/x86_64/amd/zen2/software
no software packages in tarball
other under 2023.06/software/linux/x86_64/amd/zen2
2023.06/scripts/gpu_support/nvidia/install_cuda_and_libraries.sh
```

That's not the artefacts list I expected :D Yes, this script should be reinstalled. But it should at the very least also install the `patchelf` I listed in the EasyStack file. Looking at the logs:

```
== Running parse hook for patchelf-0.18.0-GCCcore-13.2.0.eb...
== Running parse hook for GCCcore-13.2.0.eb...
== patchelf/0.18.0-GCCcore-13.2.0 is already installed (module found), skipping
== No easyconfigs left to be built.
== Build succeeded for 0 out of 0
```

Ah, that explains! Someone already installed my favorite dummy package that I like to try ;-)

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2

**Comment by casparvl** [comment]
https://github.com/EESSI/software-layer/pull/790/#issuecomment-2428351321 proves the adaptations to the script works. I removed util-linux, since that was just to prove the functionality. https://github.com/EESSI/software-layer/pull/790/#issuecomment-2428432695 is to actually deploy the slightly modified `install_cuda_and_libraries.sh`

File: EESSI-install-software.sh [hunk 2]
```diff
@@ -278,6 +262,52 @@ if command_exists "nvidia-smi"; then
...
+if [ ! -z ${EESSI_ACCELERATOR_TARGET} ]; then
+    CPU_ONLY_MODULES_PATH=$(echo $EASYBUILD_INSTALLPATH | sed "s@/accel/${EESSI_ACCELERATOR_TARGET}@@g")/modules/all
+    if [ -d ${CPU_ONLY_MODULES_PATH} ]; then
+        module use ${CPU_ONLY_MODULES_PATH}
+    else
+        fatal_error "Derived path to CPU-only modules does not exist: ${CPU_ONLY_MODULES_PATH}"
+    fi
+fi
+
+# If in dev.eessi.io, allow building on top of softw
```
Comment (ocaisa) [inline EESSI-install-software.sh:294]:
```suggestion
# If in dev.eessi.io, allow building on top of software.eessi.io
```

File: EESSI-install-software.sh [hunk 3]
```diff
@@ -278,6 +262,52 @@ if command_exists "nvidia-smi"; then
...
+if [[ "${EESSI_CVMFS_REPO}" == /cvmfs/dev.eessi.io ]]; then
+    module use /cvmfs/software.eessi.io/versions/$EESSI_VERSION/software/${EESSI_OS_TYPE}/${EESSI_SOFTWARE_SUBDIR_OVERRIDE}/modules/all
```
Comment (ocaisa) [inline EESSI-install-software.sh:296]: Would be good to make a comment that explains the expected precedence in `MODULEPATH`:
```
/path/to/dev.eessi.io/modules:/path/to/accel/modules:/path/to/software.eessi.io/modules
```

Comment (ocaisa) [reply EESSI-install-software.sh:296]: I'm not sure you have caught the case though that would define this:
* I want to install in dev.eessi.io
* I want to install for an accelerator (and require accelerator dependencies)

**Review by ocaisa — CHANGES_REQUESTED** [review]
Looks good, I think I see an issue in a particular scenario for dev.eessi.io

File: EESSI-install-software.sh (hunk 3 above)
Comment (ocaisa) [reply EESSI-install-software.sh:296]: I wonder is this what we want to support:
```
/path/to/dev.eessi.io/accel/modules:/path/to/dev.eessi.io/modules:/path/to/software.eessi.io/accel/modules:/path/to/software.eessi.io/modules
```

Comment (casparvl) [reply EESSI-install-software.sh:296]: While my pr _moved_ this section of code, I'm pretty sure it was already there, wasn't it? I'm on my phone currently, so can't easily check

Comment (casparvl) [reply EESSI-install-software.sh:296]: I'd be in favour of doing that in a separate PR, if @ocaisa agrees. This PR currently is mostly a reshuffle: some parts that can only be done after EESSI is initialized got moved down until after the EESSI module is loaded. While the pr replaced sourcing the various init scripts with loading the module, it _should_ not modify behaviour in any way. I think it's more clear of we do a big fix like that in a separate PR

Comment (ocaisa) [reply EESSI-install-software.sh:296]: I've opened https://github.com/EESSI/software-layer/issues/799 to keep track of this

**Review by ocaisa — APPROVED** [review]

**Comment by ocaisa** [comment]
Staging PR has been merged

**Pull request merged**